File: python/53_Maximum_Subarray.py
class Solution:
    # @param {integer[]} nums
    # @return {integer}
    def maxSubArray(self, nums):
        length=len(nums)
        if length<1:
            return
        maxx=res=nums[0]
        for i in xrange(1,length):
            maxx=max(maxx+nums[i],nums[i])
            res=max(res,maxx)
        return res

File: Java/53_Maximum_Subarray.java
public class Solution {
    public int maxSubArray(int[] nums) {
        if(nums==null||nums.length==0)return 0;
        int res = Integer.MIN_VALUE;
        int last = 0;
        for(int i=0;i<nums.length;i++){
            last = Math.max(last + nums[i], nums[i]);
            res = Math.max(res,last);
        }
        return res;
    }
    
    /*
    // 跟上面一种方法等价，但是少一次比较
    public int maxSubArray(int[] nums) {
        if(nums==null||nums.length==0)return 0;
        int res = nums[0];
        int last = nums[0];
        for(int i=1;i<nums.length;i++){
            last = Math.max(last + nums[i], nums[i]);
            res = Math.max(res,last);
        }
        return res;
    }
    */
}